File: apps/api/app/schemas/odd.py
"""Schemas: sistema de Odds (apostas com multiplicador)."""

import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.user import UserStateOut

Currency = Literal["silver", "gold"]


class OddOption(BaseModel):
    key: str
    label: str
    odd: float


class OddOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    title: str
    type: str
    description: str | None = None
    team_a_name: str
    team_a_odd: float
    team_b_name: str
    team_b_odd: float
    options: list[OddOption] = []
    status: str
    winner: str | None = None
    closes_at: datetime | None = None
    betting_open: bool = True
    # Agregados (preenchidos pelo serviço).
    bet_count: int = 0
    my_bets: list["OddBetOut"] = []


class OddCreate(BaseModel):
    title: str = Field(min_length=1, max_length=100)
    type: Literal["vitoria", "placar"] = "vitoria"
    description: str | None = Field(default=None, max_length=240)
    # Tipo "vitoria": confronto Time A x Time B.
    team_a_name: str | None = Field(default=None, max_length=60)
    team_a_odd: float | None = Field(default=None, ge=1.0, le=1000.0)
    team_b_name: str | None = Field(default=None, max_length=60)
    team_b_odd: float | None = Field(default=None, ge=1.0, le=1000.0)
    # Tipo "placar": 1 multiplicador comum + alternativas.
    multiplier: float | None = Field(default=None, ge=1.0, le=1000.0)
    alternatives: list[str] | None = None
    # Prazo para apostar (opcional). Após isso, ninguém mais aposta.
    closes_at: datetime | None = None


class OddUpdate(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=100)
    description: str | None = Field(default=None, max_length=240)
    team_a_name: str | None = Field(default=None, min_length=1, max_length=60)
    team_a_odd: float | None = Field(default=None, ge=1.0, le=1000.0)
    team_b_name: str | None = Field(default=None, min_length=1, max_length=60)
    team_b_odd: float | None = Field(default=None, ge=1.0, le=1000.0)
    multiplier: float | None = Field(default=None, ge=1.0, le=1000.0)
    alternatives: list[str] | None = None
    closes_at: datetime | None = None


class SettleRequest(BaseModel):
    winner: str = Field(min_length=1, max_length=20)


class PlaceBetRequest(BaseModel):
    odd_id: uuid.UUID
    selection: str = Field(min_length=1, max_length=20)
    currency: Currency = "silver"
    amount: int = Field(ge=1, le=100_000_000)


class OddBetOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    odd_id: uuid.UUID
    selection: str
    selection_label: str = ""
    currency: str
    amount: int
    odd_value: float
    status: str
    payout: int
    # Contexto da Odd (para a tela "minhas apostas" / histórico).
    odd_title: str = ""
    odd_type: str = ""
    team_a_name: str = ""
    team_b_name: str = ""
    odd_status: str = ""
    odd_winner: str | None = None


class PlaceBetResult(BaseModel):
    state: UserStateOut
    bet: OddBetOut


# Admin: resumo de uma aposta de usuário numa Odd.
class OddBetAdminOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    user_id: uuid.UUID
    team_name: str = ""
    selection: str
    currency: str
    amount: int
    odd_value: float
    status: str
    payout: int


class OddAdminDetailOut(BaseModel):
    odd: OddOut
    bets: list[OddBetAdminOut]


# Resolve a referência adiantada (OddOut.my_bets -> OddBetOut).
OddOut.model_rebuild()

File: packages/shared/src/odd.ts
// Sistema de Odds (apostas com multiplicador, criadas pelo dono).
import type { UserState } from "./user";

export type OddType = "vitoria" | "placar";
/** Chave da opção apostada: "a"/"b" (vitória) ou "opt0", "opt1"… (placar). */
export type OddSelection = string;
export type OddCurrency = "silver" | "gold";

export const ODD_TYPE_LABEL: Record<OddType, string> = {
  vitoria: "Vitória (confronto A x B)",
  placar: "Placar (alternativas)",
};

export interface OddOption {
  key: string;
  label: string;
  odd: number;
}

export interface OddBet {
  id: string;
  odd_id: string;
  selection: OddSelection;
  selection_label: string;
  currency: OddCurrency;
  amount: number;
  odd_value: number;
  status: "pending" | "won" | "lost" | "refunded";
  payout: number;
  odd_title: string;
  odd_type: string;
  team_a_name: string;
  team_b_name: string;
  odd_status: string;
  odd_winner: OddSelection | null;
}

export interface Odd {
  id: string;
  title: string;
  type: OddType;
  description: string | null;
  team_a_name: string;
  team_a_odd: number;
  team_b_name: string;
  team_b_odd: number;
  /** Opções unificadas (vitória → A/B; placar → alternativas). */
  options: OddOption[];
  status: "open" | "settled" | "cancelled";
  winner: OddSelection | null;
  closes_at: string | null;
  betting_open: boolean;
  bet_count: number;
  my_bets: OddBet[];
}

export interface PlaceBetResult {
  state: UserState;
  bet: OddBet;
}

// Admin
export interface OddBetAdmin {
  id: string;
  user_id: string;
  selection: OddSelection;
  currency: OddCurrency;
  amount: number;
  odd_value: number;
  status: string;
  payout: number;
}

export interface OddAdminDetail {
  odd: Odd;
  bets: OddBetAdmin[];
}

/** Opções de uma Odd (com fallback A/B para dados antigos sem `options`). */
export function oddOptions(o: Odd): OddOption[] {
  if (o.options && o.options.length) return o.options;
  return [
    { key: "a", label: o.team_a_name, odd: o.team_a_odd },
    { key: "b", label: o.team_b_name, odd: o.team_b_odd },
  ];
}

/** Rótulo legível de uma seleção (key) dentro de uma Odd. */
export function oddLabel(o: Odd, key: string | null): string {
  if (!key) return "";
  return oddOptions(o).find((opt) => opt.key === key)?.label ?? key;
}

/** Pagamento previsto: ceil(valor × multiplicador) — sempre arredonda p/ cima. */
export function oddPayout(amount: number, oddValue: number): number {
  return Math.ceil((Number(amount) || 0) * oddValue);
}
